src/clj/disaster/middleware/oauth2_jwt.clj

```diff
@@ -0,0 +1,197 @@
+       (group-by :kid)
+       (fmap first)
+       (fmap #(assoc {}
+                     :public-key (buddy-jwk/jwk->public-key %)
+                     :private-key (buddy-jwk/jwk->private-key %)))))
+
+(s/fdef fetch-keys
+  :args (s/cat :jwks-url ::jwks-url)
+  :ret  (s/with-gen ::key-store
+          #(s/gen #{(->> (resource "jwks.json")
+                         slurp
+                         ((fn [jwks-string] (json/read-str jwks-string :key-fn keyword)))
+                         jwks-edn->keys)})))
+
+(defn- fetch-keys
+  "Fetches the jwks from the supplied jwks-url and converts to java Keys.
+  Returns a map keyed on key-id where each value is a RSAPublicKey object"
+  [jwks-url]
+  (log/debug "Fetching keys from jwks-url" jwks-url)
+  (try  (->> jwks-url
+             slurp
+             (#(json/read-str % :key-fn keyword))
+             jwks-edn->keys)
+        (catch Exception e (do (log/error "Could not fetch jwks keys")
+                               false))))
+
+
+;; Atom to hold the public and private keys used for signature validation in memory for
+;; caching purposes. The atom holds a clojure map with kid -> key pairs. Each key is a
+;; clojure map containing a :public-key and optionally a :private-key.
+(defonce keystore
+  (atom {}))
+
+
+(defn- resolve-key
+  "Returns java.security.Key given key-fn, jwks-url and :key-type in jwt-header.
+  If no key is found refreshes"
+  [key-type jwks-url jwt-header]
+  (log/debug "Resolving key" jwt-header "from jwk cache for" jwks-url)
+  (let [key-fn (fn [] (get-in @keystore [jwks-url (:kid jwt-header) key-type]))]
+    (if-let [key (key-fn)]
+      key
+      (do (log/debug "Fetch and resolve key" jwt-header "from" jwks-url)
+          (when-let [new-keys (fetch-keys jwks-url)]
+            (swap! keystore #(update % jwks-url merge new-keys)))
+          (if-let [key (key-fn)]
+            key
+            (do
+              (log/error "Could not locate public key corresponding to jwt header's kid:" (:kid jwt-header) "for url:" jwks-url)
+              (throw (ex-info (str "Could not locate key corresponding to jwt header's kid: " (:kid jwt-header) " for url: " jwks-url)
+                              {:type :validation :cause :unknown-key}))))))))
+
+
+(s/fdef resolve-public-key
+  :args (s/cat :jwks-url    ::jwks-url
+               :jwt-header  ::jwt-header)
+  :ret  ::public-key)
+
+(def resolve-public-key
+  "Returns java.security.PublicKey given jwks-url and :kid in jwt-header.
+  If no key is found refreshes"
+  (partial resolve-key :public-key))
+
+
+(s/fdef resolve-private-key
+  :args (s/cat :jwks-url ::jwks-url
+               :jwt-header ::jwt-header)
+  :ret  ::private-key)
+
+(def resolve-private-key
+  (partial resolve-key :private-key))
+
+
+(s/fdef unsign
+  :args (s/cat :jwks-url ::jwks-url
+               :token    ::jwt)
+  :ret  ::claims)
+
+(defn- remove-bearer [token]
+  (if (and token (str/starts-with? (str/lower-case token) "bearer "))
```

Comment:
if you've already validated `token` via the above `unsign` spec, don't need to check if it's nil again here. Prob can remove the `and` entirely